grpc/grpc.go

```diff
@@ -63,21 +73,84 @@ func New(sv space.Service, fc *fuse.Controller, opts ...ServerOption) *grpcServe
 	return srv
 }
 
-// Start grpc server with provided options
+// Start grpc and api server with provided options
 func (srv *grpcServer) Start(ctx context.Context) error {
 	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", srv.opts.port))
 	if err != nil {
 		log.Error(fmt.Sprintf("failed to listen on port : %v", srv.opts.port), err)
 		return err
 	}
 
+	log.Info(fmt.Sprintf("listening on address %s", lis.Addr().String()))
+
 	srv.s = grpc.NewServer()
 	pb.RegisterSpaceApiServer(srv.s, srv)
 
+	if err = srv.startRestProxy(ctx, lis); err != nil {
+		return err
+	}
+	srv.startGrpcWebProxy()
+
+	log.Info(fmt.Sprintf("gRPC server started on Port %v", srv.opts.port))
 	srv.isStarted = true
 	srv.readyCh <- true
+	// this is a blocking function
+	return srv.s.Serve(lis)
+}
+
+func (srv *grpcServer) startRestProxy(ctx context.Context, lis net.Listener) error {
+	mux := runtime.NewServeMux()
+	opts := []grpc.DialOption{grpc.WithInsecure()}
+	if err := pb.RegisterSpaceApiHandlerFromEndpoint(ctx, mux, lis.Addr().String(), opts); err != nil {
+		log.Error("Failed to start REST server", err)
+		return err
+	}
+
+	swaggerPrefix := "/swaggerui/"
+	swaggerHandler, err := srv.getSwaggerHandler(swaggerPrefix)
+	if err != nil {
+		// QQ: Should we fail launch if error in mounting swagger docs?
+		// For now, just log a warning
+		log.Warn("Swagger UI failed to load", "err:"+err.Error())
+	}
+
+	srv.restServer = &http.Server{
+		Addr:    fmt.Sprintf(":%d", srv.opts.restProxyPort),
+		Handler: mux,
+	}
+
+	srv.restServer.Handler = cors.AllowAll().Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Debug("Incoming REST Proxy Request", "path:"+r.URL.Path, "method:"+r.Method)
+		if swaggerHandler != nil && strings.HasPrefix(r.URL.Path, swaggerPrefix) && r.Method == "GET" {
+			log.Debug("Serving swagger ui")
+			swaggerHandler.ServeHTTP(w, r)
+		} else {
+			mux.ServeHTTP(w, r)
+		}
+	}))
+
+	go func() {
+		if err := srv.restServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Error(fmt.Sprintf("REST server failed to start on port %d", srv.opts.restProxyPort), err)
+		}
+	}()
+	log.Info(fmt.Sprintf("REST server started on Port %v", srv.opts.restProxyPort))
```

Comment:
Wouldn't this log even if the server failed to start since `ListenAndServe` is in a go routine?

Reply:
Yeah, true, but if `ListenAndServe` fails we get another logged error message.
Just to make it more clearing I'm updated this log message to `REST server is starting`